src/main/java/uk/gov/ons/census/casesvc/service/CaseReceiptService.java

```diff
@@ -1,27 +1,62 @@
 package uk.gov.ons.census.casesvc.service;
 
+import static uk.gov.ons.census.casesvc.utility.QuestionnaireTypeHelper.isIndividualQuestionnaireType;
 import static uk.gov.ons.census.casesvc.utility.QuestionnaireTypeHelper.iscontinuationQuestionnaireTypes;
 
+import java.util.Optional;
 import org.springframework.stereotype.Component;
 import uk.gov.ons.census.casesvc.model.entity.Case;
 import uk.gov.ons.census.casesvc.model.entity.UacQidLink;
+import uk.gov.ons.census.casesvc.model.repository.CaseRepository;
 
 @Component
 public class CaseReceiptService {
   private final CaseService caseService;
+  private final CaseRepository caseRepository;
 
-  public CaseReceiptService(CaseService caseService) {
+  public CaseReceiptService(CaseService caseService, CaseRepository caseRepository) {
     this.caseService = caseService;
+    this.caseRepository = caseRepository;
   }
 
   public void receiptCase(UacQidLink uacQidLink) {
     Case caze = uacQidLink.getCaze();
 
     if (caze.isReceiptReceived()) return;
 
-    if (!iscontinuationQuestionnaireTypes(uacQidLink.getQid())) {
-      caze.setReceiptReceived(true);
-      caseService.saveAndEmitCaseUpdatedEvent(caze);
+    if (iscontinuationQuestionnaireTypes(uacQidLink.getQid())) return;
+
+    if (caze.getCaseType().equals("CE") && isIndividualQuestionnaireType(uacQidLink.getQid())) {
+      incrementActualResponseAndSetReceiptedIfAppropriate(caze);
+      return;
     }
+
+    caze.setReceiptReceived(true);
+    caseService.saveAndEmitCaseUpdatedEvent(caze);
+  }
+
+  private void incrementActualResponseAndSetReceiptedIfAppropriate(Case caze) {
+    /*
+      This stops the actualResponses being updated by another thread for another receipt or linking event
+    */
+    Optional<Case> oCase = caseRepository.getCaseAndLockByCaseId(caze.getCaseId());
+
+    oCase.ifPresentOrElse(
```

Comment:
**TWO** lambdas?!?! why not just throw the exception if the case isn't present, but otherwise carry on as normal? It'd be less code

Reply:
**Fair** shout!!!!